### DISCUSSION TITLE
duckdb.parser error when running example code for linking two tables

### DISCUSSION THREAD

**Discussion by maxmayer1** [moj-analytical-services/splink#1500] [Q&A]
Hi all, 

looking for your help for a seemingly simple problem. All I'm trying to do is run the example code for linking to tables provided here: https://moj-analytical-services.github.io/splink/demos/examples/duckdb/link_only.html

Installation and all look good but when running the line: 
linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)

I get an error from the duckdb parser (see full error message and code below).  Any idea where the mistake is coming from? Thank you very much for your help. 

Error message: 
duckdb.ParserException: Parser Error: syntax error at or near ";"
LINE 13: ...count_cumulative_blocks_6286df18d
        AS
        (WITH __splink__df_concat as (select * from __splink__df_concat_f7ac6fac1), 
__splink__df_blocked_data as (
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."surname" as "surname_l", "r"."surname" as "surname_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r", "l"."city" as "city_l", "r"."city" as "city_r", "l"."email" as "email_l", "r"."email" as "email_r"
            , '0' as match_key
            
            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1)
                                                   ^
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\IPython\core\interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-31-6daa6e729d35>", line 1, in <module>
    linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\linker.py", line 3156, in estimate_probability_two_random_records_match
    records = cumulative_comparisons_generated_by_blocking_rules(
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\analyse_blocking.py", line 100, in cumulative_comparisons_generated_by_blocking_rules
    cumulative_blocking_rule_count = linker._execute_sql_pipeline([concat])
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\linker.py", line 582, in _execute_sql_pipeline
    raise e
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\linker.py", line 576, in _execute_sql_pipeline
    dataframe = self._sql_to_splink_dataframe_checking_cache(
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\linker.py", line 842, in _sql_to_splink_dataframe_checking_cache
    splink_dataframe = self._execute_sql_against_backend(
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\duckdb\linker.py", line 220, in _execute_sql_against_backend
    self._log_and_run_sql_execution(sql, templated_name, physical_name)
  File "C:\Users\nb\PycharmProjects\dnb\venv\lib\site-packages\splink\linker.py", line 656, in _log_and_run_sql_execution
    raise SplinkException(
splink.exceptions.SplinkException: Error executing the following sql for table `__splink__df_count_cumulative_blocks`(__splink__df_count_cumulative_blocks_6286df18d):
        CREATE TABLE __splink__df_count_cumulative_blocks_6286df18d
        AS
        (WITH __splink__df_concat as (select * from __splink__df_concat_f7ac6fac1), 
__splink__df_blocked_data as (
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."surname" as "surname_l", "r"."surname" as "surname_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r", "l"."city" as "city_l", "r"."city" as "city_r", "l"."email" as "email_l", "r"."email" as "email_r"
            , '0' as match_key
            
            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1)
            
            where l."source_dataset" || '-__-' || l."unique_id" < r."source_dataset" || '-__-' || r."unique_id" and l."source_dataset" != r."source_dataset"
            union all
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."surname" as "surname_l", "r"."surname" as "surname_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r", "l"."city" as "city_l", "r"."city" as "city_r", "l"."email" as "email_l", "r"."email" as "email_r"
            , '1' as match_key
            
            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.surname = r.surname and levenshtein(r.dob, l.dob) &lt;= 1)
            AND NOT (coalesce((l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1), false))
            where l."source_dataset" || '-__-' || l."unique_id" < r."source_dataset" || '-__-' || r."unique_id" and l."source_dataset" != r."source_dataset"
            union all
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."surname" as "surname_l", "r"."surname" as "surname_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r", "l"."city" as "city_l", "r"."city" as "city_r", "l"."email" as "email_l", "r"."email" as "email_r"
            , '2' as match_key
            
            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.first_name = r.first_name and levenshtein(r.surname, l.surname) &lt;= 2)
            AND NOT (coalesce((l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1), false) OR coalesce((l.surname = r.surname and levenshtein(r.dob, l.dob) &lt;= 1), false))
            where l."source_dataset" || '-__-' || l."unique_id" < r."source_dataset" || '-__-' || r."unique_id" and l."source_dataset" != r."source_dataset"
            union all
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."surname" as "surname_l", "r"."surname" as "surname_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r", "l"."city" as "city_l", "r"."city" as "city_r", "l"."email" as "email_l", "r"."email" as "email_r"
            , '3' as match_key
            
            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.email = r.email)
            AND NOT (coalesce((l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1), false) OR coalesce((l.surname = r.surname and levenshtein(r.dob, l.dob) &lt;= 1), false) OR coalesce((l.first_name = r.first_name and levenshtein(r.surname, l.surname) &lt;= 2), false))
            where l."source_dataset" || '-__-' || l."unique_id" < r."source_dataset" || '-__-' || r."unique_id" and l."source_dataset" != r."source_dataset"
            ) 
        select
        count(*) as row_count,
        match_key
        from __splink__df_blocked_data
        group by match_key
        order by cast(match_key as int) asc
    )
        
Error was: Parser Error: syntax error at or near ";"
LINE 13: ...count_cumulative_blocks_6286df18d
        AS
        (WITH __splink__df_concat as (select * from __splink__df_concat_f7ac6fac1), 
__splink__df_blocked_data as (
            select
            "l"."source_dataset" as "source_dataset_l", "r"."source_dataset" as "source_dataset_r", "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", "l"."first_name" as "first_name_l", "r"."first_name" as "first_name_r", "l"."surname" as "surname_l", "r"."surname" as "surname_r", "l"."dob" as "dob_l", "r"."dob" as "dob_r", "l"."city" as "city_l", "r"."city" as "city_r", "l"."email" as "email_l", "r"."email" as "email_r"
            , '0' as match_key
            
            from __splink__df_concat as l
            inner join __splink__df_concat as r
            on
            (l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1)
                                                   ^

Code: 
from splink.duckdb.linker import DuckDBLinker
import splink.duckdb.comparison_library as cl
import splink.duckdb.comparison_template_library as ctl
import splink.duckdb.blocking_rule_library as brl
from splink.datasets import splink_datasets
import pyodbc
from sqlalchemy import create_engine

df = splink_datasets.fake_1000

df_l = df.sample(frac=0.5)
df_r = df.drop(df_l.index)

from splink.duckdb.linker import DuckDBLinker
import splink.duckdb.comparison_library as cl
import splink.duckdb.comparison_template_library as ctl

settings = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name",
        "l.surname = r.surname",
    ],
    "comparisons": [
        ctl.name_comparison("first_name",),
        ctl.name_comparison("surname"),
        ctl.date_comparison("dob", cast_strings_to_date=True),
        cl.exact_match("city", term_frequency_adjustments=True),
        ctl.email_comparison("email", include_username_fuzzy_level=False),
    ],
}

linker = DuckDBLinker([df_l, df_r], settings, input_table_aliases=["df_left", "df_right"])
deterministic_rules = [
    "l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1",
    "l.surname = r.surname and levenshtein(r.dob, l.dob) &lt;= 1",
    "l.first_name = r.first_name and levenshtein(r.surname, l.surname) &lt;= 2",
    "l.email = r.email"
]

linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)

**Comment by RobinL** [comment]
It looks like when you copied and pasted you got some html characters in your blocking rule.  You have ' &lt;=', which should be <=
